android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/ListItemProxy.java

```diff
@@ -301,6 +260,195 @@ public void call(Object data)
 		return parent;
 	}
 
+	private void addTemplateEventListeners(final TiViewProxy proxy, KrollDict template)
+	{
+		// Validate arguments.
+		if ((proxy == null) || (template == null)) {
+			return;
+		}
+
+		// Fetch events assigned to given template.
+		final KrollDict eventDictionary = template.getKrollDict(TiC.PROPERTY_EVENTS);
+		if (eventDictionary == null) {
+			return;
+		}
+
+		// Make sure we only add events listeners once for the root ListItem.
+		// Note: ListItem's view is always re-created, unlike child views which are only created once.
+		if (proxy instanceof ListItemProxy) {
+			if (this.hasAddedItemEvents) {
+				return;
+			}
+			this.hasAddedItemEvents = true;
+		}
+
+		// Add event listeners to given proxy.
+		for (KrollDict.Entry<String, Object> entry : eventDictionary.entrySet()) {
+			if (!(entry.getValue() instanceof KrollFunction)) {
+				continue;
+			}
+			final String eventName = entry.getKey();
+			final KrollFunction callback = (KrollFunction) entry.getValue();
+			final KrollObject krollObject = proxy.getKrollObject();
+			proxy.addEventListener(eventName, new KrollEventCallback() {
+				@Override
+				public void call(Object data)
+				{
+					ListItemProxy itemProxy = getListItemContainerFor(proxy);
+					if (itemProxy != null) {
+						callback.call(krollObject, new Object[] { itemProxy.handleEvent(eventName, data, false) });
+					}
+				}
+			});
+			krollObject.setHasListenersForEventType(eventName, true);
+		}
+	}
+
+	private static ListItemProxy getListItemContainerFor(TiViewProxy proxy)
+	{
+		if (proxy == null) {
+			return null;
+		} else if (proxy instanceof ListItemProxy) {
+			return (ListItemProxy) proxy;
+		}
+		return getListItemContainerFor(proxy.getParent());
+	}
+
+	public void moveChildrenTo(ListItemProxy proxy)
+	{
+		// Validate argument.
+		if (proxy == null) {
+			return;
+		}
+
+		// Copy all child view proxy properties to this list item's dictionary.
+		// Saves view's current state when scrolled offscreen so we can restore it when scrolled back in.
+		// Note: Given "proxy" shouldn't have any child proxies, but do the copy just in case.
+		copyChildPropertiesFromProxies();
+		proxy.copyChildPropertiesFromProxies();
+
+		// Remove all children from given proxy. (If recycling is working as expected, it should have none.)
+		proxy.removeAllChildren();
+
+		// Do not continue if templates don't match. Prevents given proxy from re-using this proxy's views.
+		// Note: Should never happen if ListViewAdapter.getItemViewType() returns different IDs for different templates.
+		String templateId = getTemplateId();
+		if ((templateId == null) || !templateId.equals(proxy.getTemplateId())) {
+			return;
+		}
+
+		// Do not continue if this proxy has no children.
+		// This will force given proxy to regenerate child proxies from template.
+		TiViewProxy[] childProxies = getChildren();
```

Comment:
Very minor, but saves an allocation and makes use of the `hasChildren()` method you introduced
```Java
if (!hasChildren()) {
    return;
}
final TiViewProxy[] childProxies = getChildren();
if (childProxies[0].peekView() == null) {
    return;
}
```